### PR TITLE
Performance improvements of dictionary classifiers + Multivariate MUSE classifier

### PR DESCRIPTION
**Pull request by patrickzib**
#### Reference Issues/PRs

This pull-request consists of the following changes:

1. Performance optimization for all dictionary classifiers. This yields a 2-4 fold increase in performance for all tests of TDE, BOSS, WEASEL. Partly due to saving wrapping and unwrapping pandas data frames for pipelined operations using SFA. Also adds some more numba functions
2. Provides some bugfixes to TDE, SFA (lower bounding) and WEASEL (parameters and learning the classifier)
3. Adds the WEASEL+MUSE classifier [1] for multivariate classification using dictionary classifiers to sktime.
4. Updates the documentation

[1] Schäfer, P., & Leser, U. (2018). Multivariate time series classification with WEASEL+MUSE. 3rd ECML/PKDD Workshop on AALTD.

Thank you  :)

#### PR checklist for new estimators 

- [x] I've added unit tests and made sure they pass locally. 
- [x] I've updated the existing example notebooks or provided a new one to showcase how my estimator works.
- [x] I've added myself to the list of contributors.
- [x] I've added the estimator to the online documentation.

### CODE REVIEW

**Comment by patrickzib** [comment]
@mloning Hi Markus, I saw there were a couple of changes to pre-commit. Should I format the code using „black“? I tried that locally and it changes the code layout quite significantly (not always for good, e.g. for all arrays in tests). Also there is a bug using flake8 and black when using array slices: E203 whitespace before ':' , see [1]

Black uses a different format that is not supported by flake8. Thus I had to add an ignore warning for E203 [1].

 Thanks Patrick 

[1] https://pypi.org/project/flake8-black/

**Comment by mloning** [comment]
Hi @patrickzib, alright thanks! I'm happy to merge without the pre-commit tests passing, I'm still figuring out whether they're useful or too pedantic. We're still running flake8 on azure dev ops and that passes, so all good. I'll take a look at the code later!

**Comment by mloning** [comment]
@patrickzib On the page of the linked `flake8-black` page, it says
> If you are using pre-commit configure it to call black and/or flake8 directly - you do not need flake8-black at all.

So is there another bug?

**Comment by patrickzib** [comment]
@mloning I think the pre-config calls black and then flake8 directly. Unfortunately, the same error applies as if calling the library directly ... :(

**Review by mloning — COMMENTED** [review]
Thanks @patrickzib, I had a quick looks and it looks good to me! See one question regarding the change of indexing below. 

So what shall we do with the flake8/black bug, add the `--extend-ignore` flag or leave it as it was?

Please also consider adding yourself to the CODEOWNERS file in the root dir. That way, you'd be notified automatically if someone opens a PR that affects the files you've worked on.

File: sktime/classification/dictionary_based/_boss.py
Comment (patrickzib) [reply sktime/classification/dictionary_based/_boss.py:418]: Hi @mloning, 

maybe leave the config as it was? The dependency to "black" requires reformatting all existing source codes. I am not sure, if that is intended? Most likely this makes spotting differences in pull requests to the former code even harder?

Very good spot on regarding .iloc :) I have noticed a general problem in sktime which significantly reduces runtimes and memory consumption for dictionary-methods. 

For example in BOSS we have:
BOSSEnsemble -> BOSSIndividual -> SFA. And each one requires a Pandas Dataframe as input, next converts the DataFrame using tabularize to a Numpy Array. In the end a Pandas Dataframe is returned (converted from array or dict). These conversions were very expensive for SFA and BOSSIndivudual as they consumed a significant proportion of the runtime and memory. Thus, I have changed the internal dataflow such that "tabularize" is only performed in BOSSEnsemble and numpy arrays are passed down to BOSSIndividual and SFA, saving all conversion steps inbetween. 

I noticed similar issues with PAA, but didnt address these (I suppose a lot of time from ShapeletTransform is due to the conversion steps for PAA). Maybe a more general approach would be desirable here (e.g. allowing numpy or pandas input for the feature extractors)?

Do you have an example for adding lines to CODEOWNERS? There is no entry yet...

Thank you!

Comment (mloning) [reply sktime/classification/dictionary_based/_boss.py:418]: Hi @patrickzib 
* Yes, the data container is an ongoing discussion point! I believe we made a bad choice at the start in going for nested pandas data frames instead of numpy arrays. Ideally this would be solved in the input check functions `check_X` and `check_X_y`, we're also thinking about adding support for 3d numpy arrays with shape = [n_instances, n_variables, n_timepoints] to avoid the conversion altogether for equal length series. What are your thoughts on this? How do you represent unequal length data in Python with varying number of observations across instances and/or variables?  
* CI is set up to only check the files that were changed, so this shouldn't be an issue for any new changes after pre-commit was set up. So we can ignore the failing tests on this PR. I see the point about spotting differences in PRs when additional enforcing new formatting rules. Mh, will think about removing black again, but so can you just undo the change to setup.cfg? 
* You can find more info on CODEOWNERS here: https://docs.github.com/en/github/creating-cloning-and-archiving-repositories/about-code-owners

Comment (mloning) [reply sktime/classification/dictionary_based/_boss.py:418]: Does that mean that SFA's transform method now returns numpy arrays? If so, wouldn't this break other pipelines where the next steps expects pandas data frames?

Comment (mloning) [reply sktime/classification/dictionary_based/_boss.py:418]: Or does SFA return unequal length data? I'd be very interested in hearing your opinion on the data container issue! Perhaps a quick call would be easiest?

Comment (patrickzib) [reply sktime/classification/dictionary_based/_boss.py:418]: @mloning Thank you for your answers! 

I have a 3 days old baby at home, thus it is a bit difficult to plan my time ahead of time :)

- It would be great if check_X and check_X_y would convert the data into an input format that can be used within sklearn pipelines. It would also be great if the format is compatible with numba, which supports numpy but not pandas, as of now.

- I did a kind of workaround in this pull request to avoid breaking existing functionality.
1.  For the return type of the methods: There is a constructor parameter called: `return_pandas_data_series`. When set to `False`, it will return numpy arrays. When set to `True`, it will return the default in sktime (pandas.dataframes and series). Thus, I can choose which output to use by initializing the classes, e.g. SFA(return_pandas_data_series=True) when pandas output is needed. 

2. For the input to the methods: The `fit` and `transform` methods check the input format of X and y and call `tabularize()` if it is in pandas format or keep the numpy arrays, if already in the correct format. But it would be great, if check_X and check_X_y would in the long run call tabularize directly.

- I do not really have an answer for variable lengths in numpy :(

What do you think?

Comment (mloning) [reply sktime/classification/dictionary_based/_boss.py:418]: Oh wow, congratulations! :) 

I agree, the solution isn't ideal, doing it through the input checks/conversion would be better, but don't mind it as a temporary fix. We're considering awkward-array as an alternative to numpy for unequal length data.

**Review by mloning — DISMISSED** [review]

**Comment by mloning** [comment]
Regarding the black-flake8 E203 incompatibility, see https://github.com/psf/black/issues/227

**Comment by patrickzib** [comment]
> Regarding the black-flake8 E203 incompatibility, see [psf/black#227](https://github.com/psf/black/issues/227), so I think you were right about adding it to the ignore list!

I am not familiar with formatting tools in python... Do you intend to keep black in the pipeline? Then I suppose it would be good to add the ignore line.

File: sktime/classification/dictionary_based/_boss.py
Comment (patrickzib) [reply sktime/classification/dictionary_based/_boss.py:418]: Thank you :) very exciting times...

awkward-array looks promissing! There is even an extension to numba, too: https://pypi.org/project/awkward-numba/

**Comment by mloning** [comment]
@patrickzib for now, yes, I feel that in the long run we will benefit from having a more coherent coding style I think. Having set up the CI to only check changed files, it should also work fairly well for new PRs. But I don't have a strong opinion on it, happy to reconsider if people don't like it, it does add a bit of extra setup/maintenance burden on contributors with the pre-commit workflow and so on.

**Comment by patrickzib** [comment]
@mloning @TonyBagnall not sure how to proceed with the pull request. should I push the code reformatted using `black` again?

**Comment by mloning** [comment]
Hi @patrickzib, sorry, try to run 
```
git checkout FEATURE-BRANCH
pre-commit run --files $(git diff --name-only HEAD master)
```
and sync with master. Happy to merge afterwards! 

Also note #405 in which I want to replace your kwargs with improved input checks/type coercions for handling numpy arrays.

**Comment by patrickzib** [comment]
@mloning That does not seem to work as intended. I get a lot of errors after this call, which are somewhat unrelated to the files I did change :(

```
flake8...................................................................Failed
sktime/classification/distance_based/_elastic_ensemble.py:182:17: T001 print found.
sktime/classification/distance_based/_elastic_ensemble.py:188:30: B007 Loop control variable 'test_index' not used within the loop body. If this is intended, start the name with an underscore.
sktime/classification/distance_based/_elastic_ensemble.py:194:21: T001 print found.
sktime/classification/distance_based/_elastic_ensemble.py:204:17: T001 print found.
sktime/classification/distance_based/_elastic_ensemble.py:213:13: T001 print found.
sktime/classification/distance_based/_elastic_ensemble.py:239:21: T001 print found.
sktime/classification/distance_based/_elastic_ensemble.py:247:21: T001 print found.
sktime/classification/distance_based/_elastic_ensemble.py:305:17: T001 print found.
setup.py:135:13: T001 print found.
sktime/transformers/series_as_features/dictionary_based/_sfa.py:363:33: E203 whitespace before ':'
sktime/transformers/series_as_features/segment.py:115:9: T001 print found.
sktime/forecasting/compose/_reduce.py:341:81: E203 whitespace before ':'
sktime/classification/distance_based/_proximity_forest.py:728:13: B007 Loop control variable 'index' not used within the loop body. If this is intended, start the name with an underscore.
sktime/classification/distance_based/_proximity_forest.py:1268:13: T001 print found.
sktime/transformers/series_as_features/tests/test_sfa.py:97:31: E203 whitespace before ':'
sktime/classification/frequency_based/_rise.py:155:46: E203 whitespace before ':'
sktime/classification/frequency_based/_rise.py:157:55: E203 whitespace before ':'
sktime/classification/frequency_based/_rise.py:219:46: E203 whitespace before ':'
sktime/classification/frequency_based/_rise.py:221:55: E203 whitespace before ':'
sktime/transformers/series_as_features/shapelets.py:250:21: T001 print found.
sktime/transformers/series_as_features/shapelets.py:260:21: T001 print found.
sktime/transformers/series_as_features/shapelets.py:346:51: E203 whitespace before ':'
sktime/transformers/series_as_features/shapelets.py:387:25: B007 Loop control variable 'num_cals' not used within the loop body. If this is intended, start the name with an underscore.
sktime/transformers/series_as_features/shapelets.py:396:47: E203 whitespace before ':'
sktime/transformers/series_as_features/shapelets.py:409:48: E203 whitespace before ':'
sktime/transformers/series_as_features/shapelets.py:440:21: T001 print found.
sktime/transformers/series_as_features/shapelets.py:474:25: T001 print found.
sktime/transformers/series_as_features/shapelets.py:477:89: E501 line too long (98 > 88 characters)
sktime/transformers/series_as_features/shapelets.py:485:29: T001 print found.
sktime/transformers/series_as_features/shapelets.py:503:33: T001 print found.
sktime/transformers/series_as_features/shapelets.py:536:33: T001 print found.
sktime/transformers/series_as_features/shapelets.py:583:21: T001 print found.
sktime/transformers/series_as_features/shapelets.py:738:49: E203 whitespace before ':'
sktime/transformers/series_as_features/shapelets.py:1143:13: B007 Loop control variable 'i' not used within the loop body. If this is intended, start the name with an underscore.
sktime/transformers/series_as_features/shapelets.py:1162:47: E203 whitespace before ':'

```




Also, `sktime/classification/frequency_based/_rise.py:155:46: E203 whitespace before ':'` is related to the  E203 incompatibility mentioned above.

**Comment by mloning** [comment]
Have you updated your local master with the upstream master to reflect this PR?

**Comment by mloning** [comment]
Otherwise happy to ignore pre-commit here.

**Comment by mloning** [comment]
Please also sync this branch with upstream/master prior to running the pre-commit command above

**Comment by patrickzib** [comment]
> Have you updated your local master with the upstream master to reflect this PR?

you were absolutely right about this 👍

**Comment by mloning** [comment]
Alright, perfect! Looks all good now, will merge once all tests pass! 👍

**Comment by patrickzib** [comment]
Thank you :)

**Pull request merged**
